### ISSUE TITLE
Release 0.16.0 with default values is broken !

### ISSUE THREAD

**Issue by sooslaca** [ansible/tower-operator#762]
This commit https://github.com/ansible/awx-operator/commit/eaaf55e7f0ba78e49b324e27fa7de65855b95e4f broke version 0.16.0, as it injects an extra newline.

I get:
```
"Pod \\\"awx-86f8464784-zq55x\\\" is invalid: spec.initContainers[0].image: Invalid value: \\\"quay.io/centos/centos:8 \\\": must not have leading or trailing whitespace\", \
```

Can you get a quick fix release like as 0.16.1? 

Thanks

**Comment by DySprozin** [comment]
yep! same one.

when I set the spec "project_persistence: true" in awx-demo.yml and start kubectl apply, then i get error ( kubectl -n awx describe replicasets.apps )


` replicaset-controller  (combined from similar events): Error creating: Pod "awx-demo-5f4f8797dc-ctl2j" is invalid: spec.initContainers[0].image: Invalid value: "quay.io/centos/centos:8 ": must not have leading or trailing whitespace `

in replacaset contains following wrong line:
`        image: 'quay.io/centos/centos:8 '`

**Comment by BongoEADGC6** [comment]
Seeing the same issue here. Had to manually edit the Deployment initContainer image to remove the trailing whitespace.

**Comment by laiweihwa** [comment]
Where did you edit this, @BongoEADGC6 ?

I'm seeing the same issues in 0.15.0 as well

**Comment by laiweihwa** [comment]
I've found a bit more. `roles/installer/defaults/main.yml` seems to be fine defining the container image:

```
# Use these image versions for Ansible AWX.

_image: quay.io/ansible/awx
_image_version: "{{ lookup('env', 'DEFAULT_AWX_VERSION') or 'latest' }}"
_redis_image: docker.io/redis
_redis_image_version: latest
_postgres_image: postgres
_postgres_image_version: 12
_init_container_image: quay.io/centos/centos
_init_container_image_version: 8
image_pull_policy: IfNotPresent
image_pull_secret: ''
```

It's when the ReplicaSet is created that the error is introduced. You can check your replicasets with:

`minikube kubectl -- get rs`

Here's mine:
```
lai@deploy:~/repos/awx-operator-0.16.0$ kubectl get rs
NAME                                         DESIRED   CURRENT   READY   AGE
awx-5dd57b46bf                               1         0         0       8m16s
awx-operator-controller-manager-6795f9f5f5   1         1         1       9m51s
```

When I edit it, I find the offending whitespace:
```
$ minikube kubectl -- edit rs awx-5dd57b46bf
...
        image: 'quay.io/centos/centos:8 '
```

**Comment by laiweihwa** [comment]
It looks like it happens here. I think the `newline` might create the whitespace:

`path: /opt/ansible/roles/installer/tasks/resources_configuration.yml:61\nok: [localhost] => {\"ansible_facts\": {\"_init_container_image\": \"quay.io/centos/centos:8\\n\"}, \"changed\": false}\n\r\nTASK `

But I'm unable to see how the commit eaaf55e that @sooslaca references is responsible.

**Comment by laiweihwa** [comment]
~~Ahh, yes. @sooslaca is correct.~~

I see that @shanemcd was looking to break the line up. Otherwise, the 171 characters were triggering `.yamllint`'s `max: 170` 

But when I revert to the previous version and remove one space (to get to 170), I'm still ending up with the same value in replicaset. Additionally, while `_init_container_image`, gets the whitespace, `_default_init_container_image` looks like it's interpreted properly.

**Comment by laiweihwa** [comment]
I could not find a fix other than manually editing the replicaset. Not sure how that newline/space is being introduced and I'd like to understand that.

**Comment by DySprozin** [comment]
well, my "quick-fix" (actually not fix, just downgrade to previous version with manual deploy, but you cant just download and run old awx-operator, see below)
`make undeploy`
download/unzip awx-operator version 0.15.0
` make kustomize`
` bin/kustomize build config/default > tmp.yml`
edit tmp.yml
`image: quay.io/ansible/awx-operator:latest`
==>
`image: quay.io/ansible/awx-operator:0.15.0`
apply
`kubectl apply -f tmp.yml`
...
enjoy =)

**Comment by kurokobo** [comment]
@laiweihwa 
There are several ways to write multi-line strings in YAML, starting with `|` or `>` without `-` preserves the trailing newline. This sample task for Ansible: 

```yaml
  tasks:
    - ansible.builtin.debug:
        var: multiline_samples
      vars:
        multiline_samples:
          concat_by_newline_with____trailing_newline: |
            hoge
            fuga
          concat_by_newline_without_trailing_newline: |-
            hoge
            fuga
          concat_by_space___with____trailing_newline: >
            hoge
            fuga
          concat_by_space___without_trailing_newline: >-
            hoge
            fuga
```

will outputs this strings.

```text
TASK [ansible.builtin.debug] ****************************************
ok: [localhost] => {
    "multiline_samples": {
        "concat_by_newline_with____trailing_newline": "hoge\nfuga\n",
        "concat_by_newline_without_trailing_newline": "hoge\nfuga",
        "concat_by_space___with____trailing_newline": "hoge fuga\n",
        "concat_by_space___without_trailing_newline": "hoge fuga"
    }
}
```

Same rules can be considered with Jinja2 template.

```yaml
  vars:
    default_image: hoge

  tasks:
    - ansible.builtin.debug:
        var: multiline_samples
      vars:
        multiline_samples:
          jinja2_with_newline_with____trailing_newline: |
            {{ custom_image |
               default(default_image, true) }}
          jinja2_with_newline_without_trailing_newline: |-
            {{ custom_image |
               default(default_image, true) }}
          jinja2_with_space___with____trailing_newline: >
            {{ custom_image |
               default(default_image, true) }}
          jinja2_with_space___without_trailing_newline: >-
            {{ custom_image |
               default(default_image, true) }}
```

will be:

```text
TASK [ansible.builtin.debug] ****************************************
ok: [localhost] => {
    "multiline_samples": {
        "jinja2_with_newline_with____trailing_newline": "hoge\n",
        "jinja2_with_newline_without_trailing_newline": "hoge",
        "jinja2_with_space___with____trailing_newline": "hoge\n",
        "jinja2_with_space___without_trailing_newline": "hoge"
    }
}
```

Thus, in the current implementation, regardless of the value of `_default_init_container_image`, `_init_container_image` will always contain a newline character.

**Comment by kurokobo** [comment]
@shanemcd 
Thanks for maintaining this project :)
In my understainding, `|-` or `>-` shoud be used [in this line](https://github.com/ansible/awx-operator/blob/4aeeb8db827432d9f03976bf4965924518202dc5/roles/installer/tasks/resources_configuration.yml#L63) instead of just `|`.

**Comment by sooslaca** [comment]
Sent PR https://github.com/ansible/awx-operator/pull/763 based on @kurokobo 's comment

**Comment by laiweihwa** [comment]
Did that fix things for you @sooslaca ? I'm getting the same issue after starting from scratch and then making this change manually. 

What's happening here?
```
         default(lookup('env', 'RELATED_IMAGE_AWX_INIT_CONTAINER')) |
```

**Comment by sooslaca** [comment]
I did not try, devs can test it before merge. It should work like others mentioned above too.
I think you need to build a custom docker image though to try it out and use that in the yaml as source`image:` you can't just edit in place.

**Comment by laiweihwa** [comment]
What's strange to me is that I got the same error even when using the default image instead by setting it like this:

```
- name: Set Init image URL
  set_fact:
    _init_container_image: "{{ _init_container_image }}:{{ _init_container_image_version }}"
```

**Comment by kurokobo** [comment]
Tested PR #763 and worked good in my environment. If you apply #763 manually, you will need to build your own container image for AWX Operator which includes the fix so that it can be used. A container registry is required to do this.

```bash
# After modifying resources_configuration.yml,
# build and push your own container image
BUILD_ARGS="--build-arg DEFAULT_AWX_VERSION=19.5.1" \
IMAGE_TAG_BASE=registry.example.com/awx-operator \
VERSION=0.16.0 \
make docker-build docker-push

# Then deploy modified AWX Operator using the image you built
NAMESPACE=awx \
IMAGE_TAG_BASE=registry.example.com/awx-operator \
VERSION=0.16.0 \
make deploy
```

I know it is a bit complicated, so I recommend using `0.15.0` instead until the fix is released.

```bash
# Deploy 0.15.0 of AWX Operator
git clone https://github.com/ansible/awx-operator.git
cd awx-operator
git checkout 0.15.0
export NAMESPACE=awx
make deploy
```

@sooslaca 
Thanks for the PR 👍

**Comment by shanemcd** [comment]
Hello folks! I am very sorry for not seeing this sooner. We will get https://github.com/ansible/awx-operator/pull/763 merged and a new version released ASAP.

**Comment by kurokobo** [comment]
The fix was released with `0.16.1` and worked good 🚀
